File: data structures and algorithms/java/08_017.java
import java.util.*;

/**
 * Cache that uses frequency and recency for eviction
 */
public class IntelligentCache<K, V> {
   private final int capacity;
   private final Map<K, CacheItem<K, V>> cache;
   
   /**
    * Create cache with specified capacity
    */
   public IntelligentCache(int capacity) {
      this.capacity = capacity;
      this.cache = new HashMap<>();
   }
   
   /**
    * Retrieve value and update access statistics
    */
   public V get(K key) {
      if (!cache.containsKey(key)) {
         return null;
      }
      
      // Update frequency and last access time
      CacheItem<K, V> item = cache.get(key);
      item.access();
      return item.value;
   }
   
   /**
    * Store value in cache, evicting if needed
    */
   public void put(K key, V value) {
      // Update existing item if key present
      if (cache.containsKey(key)) {
         CacheItem<K, V> item = cache.get(key);
         item.value = value;
         item.access();
         return;
      }
      
      // Evict if cache is full
      if (cache.size() >= capacity) {
         evict();
      }
      
      // Add new item
      cache.put(key, new CacheItem<>(key, value));
   }
   
   /**
    * Evict the least valuable item from cache
    */
   private void evict() {
      if (cache.isEmpty()) {
         return;
      }
      
      double minScore = Double.POSITIVE_INFINITY;
      K minKey = null;
      long currentTime = System.currentTimeMillis();
      
      // Find item with minimum value score
      for (Map.Entry<K, CacheItem<K, V>> entry : 
            cache.entrySet()) {
         CacheItem<K, V> item = entry.getValue();
         double score = item.getScore(currentTime);
         if (score < minScore) {
            minScore = score;
            minKey = entry.getKey();
         }
      }
      
      // Remove the item
      if (minKey != null) {
         cache.remove(minKey);
      }
   }
   
   /**
    * Inner class to store cache items with metadata
    */
   private static class CacheItem<K, V> {
      final K key;
      V value;
      int frequency;
      long lastAccess;
      
      CacheItem(K key, V value) {
         this.key = key;
         this.value = value;
         this.frequency = 1;
         this.lastAccess = System.currentTimeMillis();
      }
      
      /**
       * Update access statistics
       */
      void access() {
         frequency++;
         lastAccess = System.currentTimeMillis();
      }
      
      /**
       * Calculate item value based on frequency and recency
       */
      double getScore(long currentTime) {
         // Calculate recency in seconds
         double recencyScore = 1.0 / 
               (1.0 + (currentTime - lastAccess) / 1000.0);
         return frequency * recencyScore;
      }
   }
}

File: data structures and algorithms/cs/08_017.cs
using System;
using System.Collections.Generic;

public class IntelligentCache<K, V> {
    private readonly int capacity;
    private readonly Dictionary<K, CacheItem> cache;

    public IntelligentCache(int capacity) {
        this.capacity = capacity;
        this.cache = new Dictionary<K, CacheItem>();
    }

    public V Get(K key) {
        if (!cache.ContainsKey(key)) {
            return default(V);
        }

        var item = cache[key];
        item.Access();
        return item.Value;
    }

    public void Put(K key, V value) {
        if (cache.ContainsKey(key)) {
            var item = cache[key];
            item.Value = value;
            item.Access();
            return;
        }

        if (cache.Count >= capacity) {
            Evict();
        }

        cache[key] = new CacheItem(key, value);
    }

    private void Evict() {
        if (cache.Count == 0) {
            return;
        }

        double minScore = double.PositiveInfinity;
        K minKey = default(K);
        long currentTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

        foreach (var entry in cache) {
            var item = entry.Value;
            double score = item.GetScore(currentTime);
            if (score < minScore) {
                minScore = score;
                minKey = entry.Key;
            }
        }

        if (!EqualityComparer<K>.Default.Equals(minKey, default(K))) {
            cache.Remove(minKey);
        }
    }

    private class CacheItem {
        public K Key { get; }
        public V Value { get; set; }
        public int Frequency { get; private set; }
        public long LastAccess { get; private set; }

        public CacheItem(K key, V value) {
            Key = key;
            Value = value;
            Frequency = 1;
            LastAccess = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        }

        public void Access() {
            Frequency++;
            LastAccess = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        }

        public double GetScore(long currentTime) {
            double recencyScore = 1.0 / (1.0 + (currentTime - LastAccess) / 1000.0);
            return Frequency * recencyScore;
        }
    }
}